monkey/agent_plugins/exploiters/rdp/src/rdp_client.py

```diff
@@ -0,0 +1,276 @@
+import asyncio
+import logging
+from pathlib import PurePath, PureWindowsPath
+from typing import List
+
+from aardwolf import logger as aardwolf_logger
+from aardwolf.commons.factory import RDPConnectionFactory
+from aardwolf.commons.iosettings import RDPIOSettings
+from aardwolf.commons.queuedata.constants import VIDEO_FORMAT
+from aardwolf.connection import RDPConnection
+from aardwolf.extensions.RDPECLIP.clipboard import Clipboard
+from aardwolf.keyboard import VK_MODIFIERS
+from aardwolf.keyboard.layoutmanager import KeyboardLayoutManager
+from egg_timer import EggTimer
+
+from common.credentials import Credentials, NTHash, Password, Username, get_plaintext
+from common.types import NetworkPort
+from infection_monkey.i_puppet import TargetHost
+
+from .in_memory_file_provider import InMemoryFileProvider
+from .rdp_options import RDPOptions
+
+logger = logging.getLogger(__name__)
+aardwolf_logger.setLevel(logging.DEBUG)
+RDP_HANDSHAKE = "S3Cr3TtH1Ng"
+RESTART_FILE_EXPLORER_CMD = "cmd.exe /c taskkill /f /IM explorer.exe & start explorer.exe"
+VK_RETURN = "VK_RETURN"
+
+
+class RDPClient:
+    def __init__(self, options: RDPOptions):
+        self._connection: RDPConnection = None
+        self._authenticated = False
+        self._kb_layout = KeyboardLayoutManager().get_layout_by_shortname("enus")
+        self._file_provider = InMemoryFileProvider(timeout=options.file_copy_timeout)
+        self._event_loop = asyncio.get_event_loop()
+
+    def connect(
+        self, host: TargetHost, credentials: Credentials, port: NetworkPort, timeout: float
+    ):
+        try:
+            connect_task = self._event_loop.create_task(
+                self._connect_with_timeout(host, credentials, port, timeout)
+            )
+            self._event_loop.run_until_complete(connect_task)
+        except Exception as err:
+            self._event_loop.close()
+            raise err
+
+    async def _connect_with_timeout(
+        self, host: TargetHost, credentials: Credentials, port: NetworkPort, timeout: float
+    ):
+        try:
+            await asyncio.wait_for(self._connect(host, credentials, port), timeout=timeout)
+        except asyncio.TimeoutError:
+            logger.debug(f"Failed to connect to via RDP after {timeout} seconds")
+            raise
+
+    async def _connect(self, host: TargetHost, credentials: Credentials, port: NetworkPort):
```

Comment:
This function is long. Consider decomposing it into shorter functions, each of which accomplishes one thing.